### PR TITLE
Add ATAPI driver

### PR DESCRIPTION
**Pull request by MishaTY**
Adds a basic ATAPI driver.

TODO
- [x] Testing on real hardware (none of my hardware supports IDE but sata)
- [x] Test on QEMU
- [x] Test on Virtualbox
- [x] Test on Bochs

### CODE REVIEW

File: source/Cosmos.HAL2/BlockDevice/IDE.cs
Comment (MishaTY) [inline source/Cosmos.HAL2/BlockDevice/IDE.cs:43]: Should we add the ATAPI to the BlockDevices array?

**Review by charlesbetros — CHANGES_REQUESTED** [review]
Please add the comments back. Especially when a type is public

**Comment by MishaTY** [comment]
I can't test on real hardware because the dvd drive nor the hard drive is not showing up in the Blockdevice.Devices

**Comment by Arawn-Davies** [comment]
@charlesbetros have you tested on real hardware or can I do it? been meaning to set up a real debugging host and this seems perfect for it!

**Comment by charlesbetros** [comment]
I haven't tested real hardware. Not too concerned with that though. Feel free to test it.

I would like Atapio and Atapi to both inherit the ATA base class though before merging. They share a lot of code and constants and IO ports. The shared bits should be in the base class.

**Comment by Arawn-Davies** [comment]
@charlesbetros ATAPI and AtaPio are both currently inheriting BlockDevice which is what at the end of the day the System2.FatFileSystem methods call to read specific sectors. C# doesn't have Multiple Inheritance like Perl so I'm not entirely sure whether to inherit Core.ATA, which is just an IOGroup with channel and controller stuff (Primary/Secondary,Master/Slave) with similar polling Wait() methods, or just keep as is

Both do have an ATA IOGroup instance of an object in each class, which I suppose is how the inheritance is done atm. I think for future USB XHCI block devices, best leave ATA code out of that can of worms 😂 :)

**Comment by Arawn-Davies** [comment]
Personally I consider this ready to merge, but of course only when everyone's ok :)
This will need some TestRunner tests being done, but that could be done in a separate PR

**Comment by Arawn-Davies** [comment]
Personally I consider this ready to merge, but of course only when everyone's ok :)
This will need some TestRunner tests being done, but that could be done in a separate PR

**Comment by charlesbetros** [comment]
Sounds good. Merge away if it looks good.

**Review by Arawn-Davies — APPROVED** [review]
XML comments have been added back in the latest commit. All good to go!

**Comment by Arawn-Davies** [comment]
> Sounds good. Merge away if it looks good.
think you need to finish your review, can't merge :)

**Review by charlesbetros — APPROVED** [review]

**Comment by MishaTY** [comment]
That was github, did not mean to merge.

**Comment by quajak** [comment]
Was there something left to do?

**Comment by MishaTY** [comment]
Never mind, I thought that github dismissed the reviews, but @charlesbetros did.

**Pull request merged**
